### PR TITLE
OCPCLOUD-2256,OCPCLOUD-2255: refactor: remove cluster-api-operator, rework assets/manifests

### PR DESCRIPTION
**Pull request by damdo**
This PR implements the operator part of the work defined in the [Installing Cluster API components in OpenShift](https://github.com/openshift/enhancements/blob/master/enhancements/cluster-api/installing-cluster-api-components-in-ocp.md) enhancement.

Notably it:
- removes the cluster-api-operator from the worklow systems responsible for installing CAPI components in OCP
- introduces a new controller as a replacement for installing the CAPI components in the cluster
- reworks the assets/manifests generation for the provider forks to happen in the provider repositories rather than in this operator repository 

The Merge of this PR depends on:
- [x] merge of https://github.com/openshift/cluster-api/pull/179
- [x] merge of https://github.com/openshift/cluster-api-provider-aws/pull/471
- [x] merge of https://github.com/openshift/cluster-api-provider-gcp/pull/209
- [x] merge of https://github.com/openshift/cluster-api-provider-ibmcloud/pull/64
- [x] merge of any other providers.. PRs TO be ADDED
- [x] (TBD) update of openshfit/library-go now that the [AggregationRule support PR](https://github.com/openshift/library-go/pull/1623) merged

### CODE REVIEW

**Comment by damdo** [comment]
cc. @mdbooth

**Comment by damdo** [comment]
The `assets` job is expected to fail as there is no asset generation make target as no generation happening in this repo anymore.
I'll remove it from the presubmits once this and the relevant PRs are merged.

File: manifests-gen/main.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+)
+
+var basePath = flag.String("base-path", "", "path to the root of the provider's repository")
+var manifestsPath = flag.String("manifests-path", "", "path to the desired directory where to output the generated manifests")
+var providerName = flag.String("provider-name", "", "name of the provider")
+var providerType = flag.String("provider-type", "", "type of the provider")
+var providerVersion = flag.String("provider-version", "", "version of the provider")
+var projDir string
```
Comment (JoelSpeed) [inline manifests-gen/main.go:24]: Why not declare as a `var ( )` block? Do they even need to be declared as globally scoped?

File: manifests-gen/main.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	projDir = path.Join(*basePath)
+
+	p := provider{
+		Name: *providerName,
+		// TODO: improve validation
+		PType:   clusterctlv1.ProviderType(*providerType),
```
Comment (JoelSpeed) [inline manifests-gen/main.go:52]: Don't use abbreviations

File: manifests-gen/providers.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	components repository.Components
+	metadata   []byte
+}
+
+// loadComponents loads components from the given provider.
+func (p *provider) loadComponents() error {
+	// Create new clusterctl config client
+	configClient, err := configclient.New("")
+	if err != nil {
+		return err
```
Comment (JoelSpeed) [inline manifests-gen/providers.go:40]: Should wrap errors with additional context

File: manifests-gen/providers.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	return strings.ReplaceAll(strings.ToLower(string(p.PType)), "provider", "")
+}
+
+func (p *provider) writeProviderComponentsConfigmap(fileName string, objs []unstructured.Unstructured) error {
+	combined, err := utilyaml.FromUnstructured(objs)
+	if err != nil {
+		return err
+	}
+
+	annotations := openshifAnnotations
```
Comment (JoelSpeed) [inline manifests-gen/providers.go:93]: Typo

File: manifests-gen/providers.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	cmYaml, err := yaml.Marshal(cm)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(path.Join(*manifestsPath, fileName), ensureNewLine(cmYaml), 0600)
+}
+
+func importProvider(p provider) error {
+	fmt.Printf("Processing provider %s\n", p.Name)
```
Comment (JoelSpeed) [inline manifests-gen/providers.go:126]: Debug ?

**Review by JoelSpeed — COMMENTED** [review]
Please split vendor updates from code changes, very hard to review right now

File: manifests-gen/main.go (hunk 2 above)
Comment (damdo) [reply manifests-gen/main.go:52]: I didn't write any of this code, this is all code motion from https://github.com/openshift/cluster-capi-operator/blob/main/hack/assets/

File: manifests-gen/providers.go (hunk 3 above)
Comment (damdo) [reply manifests-gen/providers.go:40]: I didn't write any of this code, this is all code motion from https://github.com/openshift/cluster-capi-operator/blob/main/hack/assets/

Comment (damdo) [reply manifests-gen/providers.go:40]: Same as https://github.com/openshift/cluster-capi-operator/pull/121#discussion_r1405973164

File: manifests-gen/providers.go (hunk 4 above)
Comment (damdo) [reply manifests-gen/providers.go:93]: Same as https://github.com/openshift/cluster-capi-operator/pull/121#discussion_r1405973164

File: manifests-gen/providers.go (hunk 5 above)
Comment (damdo) [reply manifests-gen/providers.go:126]: Same as https://github.com/openshift/cluster-capi-operator/pull/121#discussion_r1405973164

**Comment by damdo** [comment]
I can split out the vendor, I squashed everything because it was quite hard to keep it rebased on top of the changes coming from main with so many stacked (+20) commits.

File: manifests-gen/providers.go (hunk 4 above)
Comment (damdo) [reply manifests-gen/providers.go:93]: Anyway, I can fix the typo

File: manifests-gen/providers.go (hunk 5 above)
Comment (damdo) [reply manifests-gen/providers.go:126]: As I read it, I think that given this is used as a CLI tool, I think the author had the intent for this to print to stdout the processing of the various components. I'm fine to change it to a log.Print if you prefer that though.

File: manifests-gen/providers.go (hunk 3 above)
Comment (damdo) [reply manifests-gen/providers.go:40]: Can add the wrapping 👍

**Comment by damdo** [comment]
@JoelSpeed modules vendoring is now separated into a different commit.
I'll squash it once approved.

File: manifests-gen/providers.go (hunk 5 above)
Comment (JoelSpeed) [reply manifests-gen/providers.go:126]: Yeah i think change these to log messages rather than `fmt.Printf`

**Review by JoelSpeed — COMMENTED** [review]
The stuff that's left is pre-existing and minor, happy to merge this as is
/lgtm

**Comment by damdo** [comment]
A new CAPG image built with RHEL9 is out, let's try one more time.
/retest

**Comment by mdbooth** [comment]
@damdo How important is this failing test? Any chance we can override it?

**Comment by damdo** [comment]
No need to override it, it is optional @mdbooth 
It looks like it failed again because the image used for CAPG in the last [test run](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-capi-operator/121/pull-ci-openshift-cluster-capi-operator-main-e2e-gcp-ovn-techpreview/1729864671629938688) `registry.build02.ci.openshift.org/ci-op-dhq91l6j/stable@sha256:44bceef247e1dad1b234e850f9509e15b37cf48387012969c0b09d5b96bfc4a6` is still rhel8:
```
[root@edf8af08eb9e /]# cat /etc/redhat-release
Red Hat Enterprise Linux release 8.6 (Ootpa)
```
hence failing with [the known OCP rhel8->rhel9 wide issue](https://redhat-internal.slack.com/archives/CB95J6R4N/p1701163468934139)

**Pull request merged**